builtin/logical/transit/lock_manager.go

```diff
@@ -0,0 +1,337 @@
+		defer lm.UnlockPolicy(name, shared)
+		return
+	}
+
+	if p == nil {
+		if !upsert {
+			defer lm.UnlockPolicy(name, shared)
+			return
+		}
+
+		// Get an exlusive lock; on success, check again to ensure that no
+		// policy exists. Note that if we are using a cache we will already be
+		// serializing this entire code path and it's currently the only one
+		// that generates policies, so we don't need to check the cache here;
+		// simply checking the disk again is sufficient.
+		lm.UnlockPolicy(name, shared)
+		lockType = exclusive
+		lm.LockPolicy(name, exclusive)
+
+		p, err = lm.getStoredPolicy(storage, name)
+		if err != nil {
+			defer lm.UnlockPolicy(name, exclusive)
+			return
+		}
+		if p != nil {
+			return
+		}
+
+		upserted = true
+
+		p = &Policy{
+			Name:       name,
+			CipherMode: "aes-gcm",
+			Derived:    derived,
+		}
+		if derived {
+			p.KDFMode = kdfMode
+		}
+
+		err = p.rotate(storage)
+		if err != nil {
+			defer lm.UnlockPolicy(name, exclusive)
+			p = nil
+		}
+
+		// We don't need to worry about upgrading since it will be a new policy
+		return
+	}
+
+	if p.needsUpgrade() {
+		lm.UnlockPolicy(name, shared)
+		lockType = exclusive
+		lm.LockPolicy(name, exclusive)
+
+		// Reload the policy with the write lock to ensure we still need the upgrade
+		p, err = lm.getStoredPolicy(storage, name)
+		if err != nil {
+			defer lm.UnlockPolicy(name, exclusive)
+			return
+		}
+		if p == nil {
+			defer lm.UnlockPolicy(name, exclusive)
+			err = fmt.Errorf("error reloading policy for upgrade")
+			return
+		}
+
+		if !p.needsUpgrade() {
+			// Already happened, return the newly loaded policy
+			return
+		}
+
+		err = p.upgrade(storage)
+		if err != nil {
+			defer lm.UnlockPolicy(name, exclusive)
+		}
+	}
+
+	return
+}
+
+func (lm *lockManager) DeletePolicy(storage logical.Storage, name string) error {
+	lm.LockAll(name)
+	defer lm.UnlockAll(name)
+
+	var p *Policy
+	var err error
+
+	if lm.CacheActive() {
+		p = lm.cache[name]
+		if p == nil {
+			return fmt.Errorf("could not delete policy; not found")
+		}
+	} else {
+		p, err = lm.getStoredPolicy(storage, name)
+		if err != nil {
+			return err
+		}
+		if p == nil {
+			return fmt.Errorf("could not delete policy; not found")
+		}
+	}
+
+	if !p.DeletionAllowed {
+		return fmt.Errorf("deletion is not allowed for this policy")
+	}
+
+	err = storage.Delete("policy/" + name)
+	if err != nil {
+		return fmt.Errorf("error deleting policy %s: %s", name, err)
+	}
+
+	err = storage.Delete("archive/" + name)
+	if err != nil {
+		return fmt.Errorf("error deleting archive %s: %s", name, err)
+	}
+
+	if lm.CacheActive() {
+		delete(lm.cache, name)
+	}
+
+	return nil
+}
+
+// When this function returns it's the responsibility of the caller to call UnlockAll if err is not nil
+func (lm *lockManager) RefreshPolicy(storage logical.Storage, name string) (p *Policy, err error) {
+	lm.LockPolicy(name, exclusive)
+
+	if lm.CacheActive() {
+		p = lm.cache[name]
+		if p != nil {
+			return
```

Comment:
Only when it's read from disk, because it will be upgraded before it ever makes it into the cache.